File: Huffman.scala
object Huffman {

  sealed trait Tree[+B] {
    def :^:[B](value: Tree[B]): Node[B] = Node(this,value)
  }
  case class Leaf[B](el: B) extends Tree[B]
  case class Node[B](tr1: Tree[B], tr2: Tree[B]) extends Tree[B]

  sealed trait With[A,B] {
    def upper (): A
    def satelite (): B
    def ==(That:With[A,B]): Boolean
    def <=(That:With[Ordered[A],B]): Boolean
  }

  case class WithAB[A,B](a: A, b: B)extends With[A,B]{

    override def upper(): A = this.a
    override def satelite(): B = this.b
    override def ==(That: With[A, B]): Boolean = this.upper() == That.upper()
    override def <=(That: With[Ordered[A], B]): Boolean = That.upper() <= this.upper()
  }

  def collapse [T](Collapsable: Iterable[List[T]]): T
  def twoTails [T](list: List[T]): List[T] = list.tail.tail

  def frequency[char](cArr: List[char]): List[WithAB[Int,char]] = cArr match {
    case Nil => List()
    case () => List()
    case c => val s = c.sorted.groupBy(x => x).values; List(WithAB(s.map(_.length).head, collapse(s))) ++ frequency(s.tail.toList.flatten)
  }


  def huffman [char](withAB: List[WithAB[Int,char]]): Tree[char] = withAB match{
    case Nil => null
    case x :: () => Leaf(x.satelite())
    case x :: y :: () => Leaf(x.satelite()) :^: Leaf(y.satelite())
    case c =>  val x = c.sorted.map(x => Leaf(x.satelite())).take(2); (x.head :^: x.last) :^: huffman(twoTails(c))
  }

  trait Bit{
    def value():Int
    def show():Char
  }

  case class I() extends Bit {
    override def value(): Int = 1
    override def show(): Char = 'I'
  }

  case class O() extends Bit {
    override def value(): Int = 0
    override def show(): Char = 'O'
  }

  def decode [char](tree: Tree[char], bits: List[Bit]): List[char] = (tree, bits) match {
    case (t, Nil) => List()
    case (t, b) =>

      def letter(tr: Tree[char],lbit: List[Bit]): (List[Bit], char) = (tr,lbit) match {
        case (Leaf(a),x) => (x,a)
        case ((_ :^: trR),(I::x)) => letter(trR,x)
        case ((trL :^: _),(O::x)) => letter(trL,x)
      }

      val (xs,l) = letter(t,b);

      l:: decode(t,xs)
  }
}

/*
  def codes [char](tree: Tree[char]):List[(char, List[Bit])] = {
    val b = List()
    tree match {
      case Leaf(x) => List((x,List[]))
      case (x :^: y) =>
        val exO = (e:char) => (e,codes(x).flatMap(v => O::v._2))
        val exI = (e:char) => (e,codes(x).flatMap(v => I::v._2))

    }
  }
*/

  /*def encode [char] (tree: Tree[Ordering[char]],cArr: List[char]): List[Bit] = (tree, cArr) match {
    case (_, Nil) => List()
    case (t, (c::cs)) =>
      def checkCode(): Unit ={

      }
  }*/

/*
sealed trait Bit extends Enumeration {
  type I(1)
  type O(0)
}

object Bit extends Bit {
  override type I = "I"
  override type O = "O"
}

def codes [char](tree: Tree[char]):List[(char,List[Bit.type ])] = tree match{
  case Leaf(x) => List((x,List[]))
  case (x :^: y) => codes(x).map((a:char,b:List[Bit]) => (a, O::b)  )
}

*/

File: Huffman.hs

module Huffman
where
import Satellite
import Tree
import Data.List
--Mario Tsatsev s1028415
-- -------------------------------------------------------------------------------
-- Warm-up: constructing a frequency table.

frequencies  ::  (Ord char) => [char] -> [With Int char]
frequencies [] = []
frequencies c = let d = sort(c) in [ head (map length (group (d))) :- collapse(group (d)) ] ++ frequencies (concat (tail (group d)))
 where
  collapse a = head(head a)
-- -------------------------------------------------------------------------------

-- Constructing a Huffman tree.

huffman :: [With Int char] -> Tree char
huffman [a]   = Leaf (satellite a)
huffman [a,b] = Leaf (satellite a) :^: Leaf (satellite b)
huffman a = ((\x ->  head x :^: last x) (take 2(map (\x -> Leaf (satellite x)) (sort a)))) :^: huffman (tail (tail a))
--well technically it gives a bit different result than the one showed in the assignment :/
-- -------------------------------------------------------------------------------

-- Encoding ASCII text.

data Bit = O | I
  deriving (Show, Eq, Ord)

encode :: (Eq char) => Tree char -> [char] -> [Bit]
encode _ [] = []
encode t (c:cs) = (checkCode c (codes t)) ++ encode t cs
  where
    checkCode x ((first,second):cs) = if x == first then second else checkCode x cs
--encode t c =  concat(map snd (codes t))
--this was the attempt that "worked" with my huffman tree.

codes :: Tree char -> [(char, [Bit])]
codes (Leaf x) = [(x,[])]
codes (x :^: y) = (map (\(a, b) -> (a, O:b)) (codes x)) ++ (map (\(c, b) -> (c, I:b)) (codes y))
-- -------------------------------------------------------------------------------

--Decoding a Huffman binary.

decode :: Tree char -> [Bit] -> [char]
decode t [] = []
decode t x = l:(decode t xs) where
  (xs, l) = letter t x
  letter (Leaf a) x = (x, a)
  letter (_ :^: b) (I:x) = letter b x
  letter (a :^: _) (O:x) = letter a x


-- -------------------------------------------------------------------------------

-- Some test data.

hw, why :: String
hw =
  "hello world"

-- code = huffman (frequencies hw)
-- encode code hw
-- decode code it
-- decode code it == hw

why =
  "As software becomes more and more complex, it\n\
  \is  more  and  more important to structure it\n\
  \well.  Well-structured  software  is  easy to\n\
  \write,   easy   to   debug,  and  provides  a\n\
  \collection  of modules that can be re-used to\n\
  \reduce future programming costs. Conventional\n\
  \languages place a conceptual limit on the way\n\
  \problems   can   be  modularised.  Functional\n\
  \languages  push  those  limits  back. In this\n\
  \paper we show that two features of functional\n\
  \languages    in    particular,   higher-order\n\
  \functions and lazy evaluation, can contribute\n\
  \greatly  to  modularity.  Since modularity is\n\
  \the key to successful programming, functional\n\
  \languages  are  vitally important to the real\n\
  \world."

-- code = huffman (frequencies why)
-- encode code why
-- decode code it
-- decode code it == why
